hoomd/mpcd/RejectionVirtualParticleFiller.h

```diff
@@ -0,0 +1,175 @@
+
+namespace mpcd
+{
+
+//! Adds virtual particles to MPCD particle data for "sphere" geometry
+/*!
+ * <detailed description needed>
+*/
+
+template<class Geometry>
+class PYBIND11_EXPORT RejectionVirtualParticleFiller : public mpcd::VirtualParticleFiller
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param sysdata - MPCD system data
+         * \param density - number density of MPCD solvent
+         * \param type - particle fill type
+         * \param T - Temperature
+         * \param seed - seed for PRNG
+         * \param geom - confinement geometry
+         */
+        RejectionVirtualParticleFiller(std::shared_ptr<mpcd::SystemData> sysdata,
+                                       Scalar density,
+                                       unsigned int type,
+                                       std::shared_ptr<::Variant> T,
+                                       unsigned int seed,
+                                       std::shared_ptr<const Geometry> geom)
+        : mpcd::VirtualParticleFiller(sysdata, density, type, T, seed), m_geom(geom)
+        {
+        m_exec_conf->msg->notice(5) << "constructing MPCD RejectionVirtualParticleFiller" << std::endl;
+        }
+
+        //! Destructor
+        ~RejectionVirtualParticleFiller()
+        {
+        m_exec_conf->msg->notice(5) << "Destroying MPCD RejectionVirtualParticleFiller" << std::endl;
+        }
+
+        //! Get the streaming geometry
+        std::shared_ptr<const Geometry> getGeometry() const
+            {
+            return m_geom;
+            }
+
+        //! Set the streaming geometry
+        void setGeometry(std::shared_ptr<const Geometry> geom)
+            {
+            m_geom = geom;
+            }
+
+        //! Fill the particles outside the confinement
+        void fill(unsigned int timestep);
+
+    protected:
+        std::shared_ptr<const Geometry> m_geom;
+
+    };
+
+
+template<class Geometry>
+void RejectionVirtualParticleFiller<Geometry>::fill(unsigned int timestep)
+    {
+    // Number of particles that we need to draw (constant)
+    const BoxDim& box = m_pdata->getBox();
+    const Scalar3 lo = box.getLo();
+    const Scalar3 hi = box.getHi();
+    const unsigned int m_NVirtMax = round(m_density*box.getVolume());
+
+    // Step 1: Create temporary GPUArrays to draw Particles locally using the worst case estimate for number
+    //         number of particles.
+    GPUArray<Scalar4> pos_loc(m_NVirtMax, m_exec_conf);
+    ArrayHandle<Scalar4> m_pos_loc(pos_loc, access_location::host, access_mode::readwrite);
+
+    // Step 2: Draw the particles.
+    unsigned int pidx = 0;
+    for (unsigned int i=0; i < m_NVirtMax; ++i)
+        {
+        // TODO: Currently just using the constant for SlitGeometryFiller which needs to be changed
+        // NOTE: particle tags neglected for the RNG here since we don't know them.
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::SlitGeometryFiller, m_seed, timestep);
```

Comment:
This **needs** to have an additional counter per particle, or it will generate the same random number for each (I'm not sure if you've tried to run it and observed this already).

I suggest that we actually compute the tag range before this loop with `computeFirstTag` then allow "holes" in the final set of tags (i.e., the tags will not be contiguous). We can then use this tag as a counter (after `timestep`).

We also need a unique `RejectionFiller` identifier in the `RNGIdentifiers` list.

Last, we should add a static variable to the base class that assigns a unique ID to each filler as it is created (0, 1, ...). Then, we should use that as an extra counter here to make sure there are unique random numbers if multiple fillers are attached to the system.